packages/models/src/utils/decorator/provider.ts

```diff
@@ -107,6 +116,21 @@ export const Lock =
       target,
     )
 
+// FIXME: refactor with Lock
+export const Type =
+  (script?: Partial<Script>): ClassDecorator =>
+  (target: object) =>
+    Reflect.defineMetadata(
+      ProviderKey.TypePattern,
+      (ref?: ActorRef) => {
+        const codeHash = script?.codeHash ?? ref?.params?.codeHash
+        const hashType = script?.hashType ?? ref?.params?.hashType
+        const args = script?.args ?? ref?.params?.args ?? '0x'
+        return codeHash && hashType ? { codeHash, hashType, args } : undefined
+      },
+      target,
+    )
+
```

Comment:
Isn't it better to extract the code like this?

```typescript
const Script =
  (patternType: symbol, script?: Partial<Script>): ClassDecorator =>
  (target: object) =>
    Reflect.defineMetadata(
      patternType,
      (ref?: ActorRef) => {
        const codeHash = script?.codeHash ?? ref?.params?.codeHash
        const hashType = script?.hashType ?? ref?.params?.hashType
        const args = script?.args ?? ref?.params?.args ?? '0x'
        return codeHash && hashType ? { codeHash, hashType, args } : undefined
      },
      target,
    )

export const Type = (script?: Partial<Script>): ClassDecorator => Script(ProviderKey.TypePattern, script)

export const Lock = (script?: Partial<Script>): ClassDecorator => Script(ProviderKey.LockPattern, script)
```

Reply:
That will be better, it needs to refactor to that.

Reply:
Sure, it's been marked with
```
// FIXME: refactor with Lock
```
It's not done in this PR because of the time left for demonstration